internal/sbi/producer/datapath.go

```diff
@@ -3,457 +3,481 @@ package producer
 import (
 	"fmt"
 
-	"github.com/free5gc/nas/nasMessage"
 	"github.com/free5gc/openapi/models"
 	"github.com/free5gc/pfcp"
 	"github.com/free5gc/pfcp/pfcpType"
-	"github.com/free5gc/pfcp/pfcpUdp"
 	smf_context "github.com/free5gc/smf/internal/context"
 	"github.com/free5gc/smf/internal/logger"
 	pfcp_message "github.com/free5gc/smf/internal/pfcp/message"
+	"github.com/google/uuid"
 )
 
-type PFCPState struct {
-	upf     *smf_context.UPF
-	pdrList []*smf_context.PDR
-	farList []*smf_context.FAR
-	barList []*smf_context.BAR
-	qerList []*smf_context.QER
-	urrList []*smf_context.URR
-}
+// Sends PDRs of all datapaths to UPFs and sends the status to the UE.
+// It returns after all PFCP response have been returned or timed out,
+// and before sending the N1N2MessageTransfer request if it is needed.
+func ActivatePDUSessionAtUPFs(
+	smContext *smf_context.SMContext,
+) smf_context.PFCPSessionResponseStatus {
+	// smContext has a map to all PFCPSessionContexts that need to be established,
+	// one for each involved UPF
+
+	resChan := make(chan smf_context.SendPfcpResult)
+	defer close(resChan)
+
+	defaultPath := smContext.Tunnel.DataPathPool.GetDefaultPath()
+	anUPF := defaultPath.FirstDPNode.UPF
```

Comment:
Why is SMF just trying to activate the first node of the path?
I'm not sure this change can work for the ULCL procedure.

Reply:
I totally agree after re-checking the code. I thought the file `internal/sbi/producer/ulcl_procedure.go` handles the ULCL procedure entirely, but this part of the code is actually never used.
I will remove this file and adapt above code.

Reply:
FYI: I have tested the ULCL scenario using docker compose with the config files I put in [this PR](https://github.com/free5gc/free5gc-compose/pull/110), and it worked now.